### DISCUSSION TITLE
defining several tests with generic parameters of the dut set to different sets of values

### DISCUSSION THREAD

**Discussion by jreinauld** [cocotb/cocotb#2871] [Q&A] [answered]
Hi all, I am a new user of cocotb, which I love.
I hope this question is not already answered (I googled a bit but found nothing)
How should I proceed if I want to define several tests in my python file, and the generics of my dut have different values according to the test?
Thanks,
- Julien

**Comment by ktbarrett** [comment, marked as answer]
The python file containing the cocotb tests is loaded within the simulator run, so it cannot change the generics at that point. What you can do is handle that in your Makefile by creating multiple targets with the proper SIM_ARGS generics set and TESTCASE set to the particular test you are trying to run with those generics, and recursively invoking make. Below is an example.

```python
# test.py

@cocotb.test()
async def test1(dut):
    ...

@cocotb.test()
async def test2(dut):
    ...
```

```make
# Makefile

test1:
    $(MAKE) sim SIM_ARGS='-gARG=1' MODULE=test TESTCASE=test1

test2:
    $(MAKE) sim SIM_ARGS='-gARG=2' MODULE=test TESTCASE=test2

include $(cocotb-config --makefiles)/Makefile.sim
```

Then you run the tests like `make test1` and `make test2`.

**Reply by jreinauld** [reply]
thx!

**Reply by xXIty** [reply]
From what I'm testing, if you run `make -j2 test1 test2` you get an undefined behavior. 
What is the intended way to parallelize the execution of both targets ?

**Reply by ktbarrett** [reply]
@xXIty Multiple `make` invocations in separate shells in separate directories.

**Comment by darrenxonaspace** [comment]
This is one thing that I find `cocotb-test` (package available from pip) to be very helpful for. You can make many different test cases with varying generics and call them all with just "pytest" at the top level. Using cocotb-test, you would replace your Makefile with something like this

```python
from cocotb_test.simulator import run

def test_dut_case1():
  run(
    verilog_sources=["dut.v"], # sources
    toplevel="dut",                 # top level HDL
    module="dut_cocotb",     # name of cocotb test module
    parameters={"GENERIC_1" :  1,
                          "GENERIC_2" : 0xDEADBEEF}
  )

def test_dut_case2():
  run(
    verilog_sources=["dut.v"], # sources
    toplevel="dut",                 # top level HDL
    module="dut_cocotb",     # name of cocotb test module
    parameters={"GENERIC_1" :  5,
                          "GENERIC_2" : 0x01234567}
  )

...
```

No need to create Makefiles or worry about recursively finding all the tests and calling them multiple times. Pytest will find and run all of them for you. There are also [more advanced ways](https://docs.pytest.org/en/7.1.x/example/parametrize.html) to parameterize your tests with pytest that allow you to add many parameter cases to a single test definition.

https://github.com/themperek/cocotb-test

**Reply by franhervas99** [reply]
Two questions about that.

- Can you use the `cocotb.runner.Simulator` class to do multiple testcases with different topmodules, or you have to use `cocotb_test.simulator`?
- With this thing of running the simulator inside the main function of the python module, the Makefile is not needed? I mean, if you put the arguments inside the `build` and `test` methods of  `cocotb.runner.Simulator`, you simply run the python file and it should work, right?

Thank you in  advance!

**Reply by ktbarrett** [reply]
`cocotb.runner` fulfils some of the same functionality as `cocotb_test` which is not a part of the cocotb repo, it is a separate package. But if you use `cocotb.runner` you don't need the Makefiles. You don't even need `cocotb.runner`, cocotb is actually very simple to integrate into your own build flow, the Makefiles and `cocotb.runner` aren't doing any black magic. If you are interested in that, [check out the documentation](https://docs.cocotb.org/en/stable/custom_flows.html).

**Reply by franhervas99** [reply]
Okay, I will take a look to the documentation.
Thank you!
